### PR TITLE
Refactor reader configuration loading to remove redundant code

### PR DESCRIPTION
**Pull request by djhoese**
I noticed while working on another PR that there are multiple places in Satpy that read the YAML reader configuration files. Mainly `satpy/readers/__init__.py` and `satpy/readers/yaml_reader.py`. This is prone to errors in the future especially since the two uses were already merging configs in different ways. I refactored the code to use a single function for this. Additionally I changed the `__init__` signature for the readers so they are provided a dictionary of the configuration information. This cleans up so much other code, especially the tests.

I think this should maybe be done for writers too.

 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``flake8 satpy`` 
 - [x] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
Oh and I forgot to mention, I had to move DatasetDict and get_key to `satpy/dataset.py` to avoid a circular dependency between `satpy/readers/__init__.py` and `satpy/readers/yaml_reader.py`. I've always wanted to do this move but I know it makes @mraspaud angry so I'm willing to compromise to another location even if it means `satpy/dataset/dataset_dict.py`.

**Comment by mraspaud** [comment]
Am I ever angry? ;)

Indeed, I'd like to have a separate module for the dataset dict, feels cleaner.

**Comment by mraspaud** [comment]
Sorry, looks like I introduced merge conflicts...

**Comment by djhoese** [comment]
I'll fix the merge conflicts, but what are your thoughts otherwise?

File: satpy/dataset/__init__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+
+from .dataid import DataID, DataQuery, ModifierTuple, create_filtered_query  # noqa
+from .data_dict import get_key, DatasetDict  # noqa
+from .metadata import combine_metadata  # noqa
+from .anc_vars import dataset_walker, replace_anc  # noqa
```
Comment (mraspaud) [inline satpy/dataset/__init__.py:22]: Are you sure about the relative imports? I prefer absolute imports myself.

File: satpy/dataset/data_dict.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    Args:
+        dquery (DataQuery): Query parameters to match in the `key_container`.
+        key_container (iterable): Set, list, tuple, or dict of `DataID`
+                                  keys.
+
+    Returns (list): List of keys matching the provided parameters in no
+                    specific order.
+
+    """
+    return dquery.filter_dataids(key_container)
```
Comment (mraspaud) [inline satpy/dataset/data_dict.py:44]: Maybe this function could be inlined altogether?

**Review by mraspaud — APPROVED** [review]
Good job on reorganizing everything. I have two minor comments, but LGTM.

File: satpy/dataset/__init__.py (hunk 1 above)
Comment (djhoese) [reply satpy/dataset/__init__.py:22]: What do you mean "am I sure"? They work and for a subpackage I think it makes sense. I could see avoiding imports like `from ...readers.yaml_reader import X` or some other really long parent directory `.` sequence. Here I don't mind it. Is there a standard we can lean on?

Comment (mraspaud) [reply satpy/dataset/__init__.py:22]: Ok

File: satpy/dataset/data_dict.py (hunk 2 above)
Comment (mraspaud) [reply satpy/dataset/data_dict.py:44]: Filter keys by parameter id was already there, I just replaced the content of it with a method call on the data query.

**Pull request merged**